MoE Decoder Functional

kpaigwar opened the pull request:
Decoder Block with MoE is functional.
`2025-08-08 02:42:40.125 | INFO     | models.demos.deepseek_v3.tests.test_decoder_block:test_forward_pass:254 - PCC for MoEDecoderBlock in decode mode: 0.9910413918076039  `

Key Changes
1. moe_decoder_block.py filled up with all the required ccl helpers
2. Fixed reference_forward dynamic caching when decoder is tested on any layer_num > 0
3. Cherry-picked all_gather on MoE input. [Feel free to optimize that and make it common for both shared experts and MoE]
4. Added weight config caching in test_decoder_block.py. [Not necessary to do the same way, it is for my testing convenience]




### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/16821291351) CI passes
- [x] [Deepseek TG](https://github.com/tenstorrent/tt-metal/actions/runs/17071151057) CI

barci2 commented on models/demos/deepseek_v3/tt/moe.py:
```diff
@@ -205,7 +217,10 @@ def create_runtime_output_buffers(
-        assert x.memory_config() == cfg["input_memory_config"]
+        x = ttnn.experimental.all_gather_async(x, **cfg["revert_tp"])
+
+        if x.memory_config() != cfg["input_memory_config"]:
```
don't think you need this, this is guaranteed by the all_gather

barci2 reviewed (changes requested):
Added some comments

barci2 commented on models/demos/deepseek_v3/conftest.py:
```diff
@@ -92,3 +92,8 @@ def ccl(mesh_device):
+
+
+@pytest.fixture(scope="session")
+def deepseek_cache_path():
```
let's add this in a separate PR maybe, I thought that for now we wanted to keep things cacheless

barci2 commented on models/demos/deepseek_v3/tt/decoder_block/moe_decoder_block.py:
```diff
@@ -88,20 +125,117 @@ def create_mlp_state(
+        DP_FACTOR = cfg["moe"][row_idx]["num_dispatch_devices"]
+        DP_SIZE = even_int_div(num_tokens_to_route, DP_FACTOR)
+        # Apply data parallelism only if the number of tokens per dispatch device is a multiple of the tile size
+        apply_dp = DP_SIZE % TILE_SIZE == 0
```
love the use of constants

barci2 commented on models/demos/deepseek_v3/tests/test_moe.py:
```diff
@@ -79,14 +79,13 @@ def test_forward_pass(
     )
 
     # TTNN forward pass using utility function
-    tt_input = ttnn.to_memory_config(tt_input, run_config["input_memory_config"])
```
why?

barci2 commented on models/demos/deepseek_v3/tests/test_decoder_block.py:
```diff
@@ -33,7 +34,8 @@ def hf_config(hf_config):
 def load_reference_model(hf_config, layer_idx):
     """Load the reference model for testing."""
-
+    torch.manual_seed(5)
```
if we're doing this, let's add it to every test. Also, let's maybe inline this function

kpaigwar replied:
Sure, we should set all over.

kpaigwar commented:
> Added some comments
Addressed, please review again

barci2 commented on models/demos/deepseek_v3/tests/test_decoder_block.py:
```diff
@@ -86,7 +79,11 @@ def test_forward_pass(
     if module_path is None:
-        reference_model = load_reference_model(hf_config, reference_layer_idx)
+        # This is needed to make the reference model weights initialization deterministic
+        torch.use_deterministic_algorithms(True)
```
maybe we can make this a fixture? Can help you with that if needed.

barci2 reviewed (approved):
lgtm

kpaigwar committed:
resolve conflicts

weight caching

tests passed

ttnn.tile_size and mem_cfg addrressd

use reset_seeds fixture

move from config_dataclasss

remove caching

remove caching

increase timeout

reduce timeout

tt-rkim commented on .github/workflows/tg-deepseek-tests-impl.yaml:
```diff
@@ -27,7 +27,7 @@ jobs:
       matrix:
         test-group: [
-          { name: "Galaxy DeepSeek tests", arch: wormhole_b0, model: deepseek, timeout: 60, owner_id: U03HY7MK4BT}, # Mark O'Connor
+          { name: "Galaxy DeepSeek tests", arch: wormhole_b0, model: deepseek, timeout: 90, owner_id: U03HY7MK4BT}, # Mark O'Connor
```
90 minutes?

kpaigwar committed:
rebase

kpaigwar committed:
revert to quantize_fp8_blockwise

kpaigwar committed:
fix typo
